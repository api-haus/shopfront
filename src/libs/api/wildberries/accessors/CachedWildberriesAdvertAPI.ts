import type { MongoClient } from 'mongodb';

import { BaseCachedAPI, CACHE_DB } from '../../../BaseCachedAPI.js';
import type { WildberriesAdvertAPI } from '../clients/WildberriesAdvertAPI.js';
import type { IIdAndDates } from '../params/IDAndDatesArg.js';
import type { IWBPromotionFullStats } from '../types/IWBPromotionFullStats.js';

const splitIdsAndDates = (args: IIdAndDates): IIdAndDates[] => {
  const splitArgs: IIdAndDates[] = [];
  const { dates = [], id } = args;

  for (const date of dates) {
    splitArgs.push({
      id,
      dates: [date],
    });
  }

  return splitArgs;
};

export class CachedWildberriesAdvertAPI extends BaseCachedAPI {
  constructor(protected inner: WildberriesAdvertAPI, protected mongo: MongoClient) {
    super(mongo);

    this.setArgsSplitter<IIdAndDates>(
      this.promotionFullStats.name,
      splitIdsAndDates,
    );
    this.setDataFetcher<IIdAndDates, IWBPromotionFullStats[]>(
      this.promotionFullStats.name,
      args => this.inner.promotionFullStats([args]),
    );

    void this.ensureIndexes();
  }

  async ensureIndexes() {
    await this.mongo.db(CACHE_DB).collection(this.promotionFullStats.name)
      .createIndex(
        {
          key: 1,
        },
        {
          unique: true,
          sparse: false,
        },
      );
  }

  async promotionFullStats(idsAndDates: IIdAndDates[]): Promise<IWBPromotionFullStats[]> {
    const allResults = await this.cachingCall<IIdAndDates, IWBPromotionFullStats[]>(
      this.promotionFullStats.name,
      idsAndDates,
    );

    return allResults.flat();
  }
}
