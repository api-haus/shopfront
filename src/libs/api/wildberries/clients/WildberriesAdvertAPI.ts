import { constants } from 'node:http2';

import Bottleneck from 'bottleneck';
import { getGlobalDispatcher, interceptors } from 'undici';

import config from '../../../../config/index.js';
import { TokenizedAPIClient } from '../../clients/TokenizedAPIClient.js';
import { retryHandler } from '../../common/RetryHandler.js';
import type { IIdAndDates } from '../params/IDAndDatesArg.js';
import type { IPromotionAdvertsInfo } from '../types/IPromotionAdvertsInfo.js';
import type { IWBAdvertList } from '../types/IWBAdvertList.js';
import type { IWBPromotionFullStats } from '../types/IWBPromotionFullStats.js';

export const CPrefixAdvertAPIV1 = 'https://advert-api.wildberries.ru/adv/v1';
export const CPrefixAdvertAPIV2 = 'https://advert-api.wildberries.ru/adv/v2';

export class WildberriesAdvertAPI extends TokenizedAPIClient {
  bottleneck = new Bottleneck({
    maxConcurrent: 1,
    minTime: 5000,
  });

  dispatcher = getGlobalDispatcher()
    .compose(interceptors.retry({
      ...config.wildberries.api.retryOptions,
      retry: retryHandler,
      errorCodes: [
        'ECONNRESET',
        'ECONNREFUSED',
        'ENOTFOUND',
        'ENETDOWN',
        'ENETUNREACH',
        'EHOSTDOWN',
        'EHOSTUNREACH',
        'EPIPE',
        'UND_ERR_SOCKET',
        'UND_ERR_REQ_CONTENT_LENGTH_MISMATCH',
      ],
      methods: [
        'GET',
        'HEAD',
        'POST',
        'OPTIONS',
        'PUT',
        'DELETE',
        'TRACE',
      ],
      statusCodes: [
        500,
        502,
        503,
        504,
        429,
      ],
    }));

  async groupedPromotionCounts(): Promise<{ adverts: IWBAdvertList[] }> {
    return this.rateLimitedCall<{ adverts: IWBAdvertList[] }>(
      {
        maxConcurrent: 1,
        minTime: 1000 / 5 + 100, // max 5 rps
      },
      `${CPrefixAdvertAPIV1}/promotion/count`,
      {
        method: 'GET',
      },
    );
  }

  async promotionFullStats(idsAndDates: IIdAndDates[]): Promise<IWBPromotionFullStats[]> {
    return this.rateLimitedCall<IWBPromotionFullStats[]>(
      {
        maxConcurrent: 1,
        minTime: 1000 / 5 + 100, // max 5 rps
      },
      `${CPrefixAdvertAPIV2}/fullstats`,
      {
        method: 'POST',
        headers: {
          [constants.HTTP2_HEADER_CONTENT_TYPE]: 'application/json',
        },
        body: JSON.stringify(idsAndDates),
      },
    ).catch((err: unknown) => {
      if ((err as Error).message.includes('нет кампаний')) {
        return [];
      }
      throw err;
    });
  }

  async promotionAdvertsInfo(advertIds: number[]): Promise<IPromotionAdvertsInfo> {
    return this.rateLimitedCall<IPromotionAdvertsInfo>(
      {
        maxConcurrent: 1,
        minTime: 1000 / 5 + 100, // max 5 rps
      },
      `${CPrefixAdvertAPIV1}/promotion/adverts`,
      {
        method: 'POST',
        headers: {
          [constants.HTTP2_HEADER_CONTENT_TYPE]: 'application/json',
        },
        body: JSON.stringify(advertIds),
      },
    );
  }
}
