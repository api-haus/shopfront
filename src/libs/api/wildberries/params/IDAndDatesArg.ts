import { toDateString } from '../../../date/toDateString.js';
import type { TDateAsString } from '../../scalars/TDateAsString.js';

export interface IIdAndDates {
  id: number;
  dates: TDateAsString[]; // ['2014-01-02']
}

export class IDAndDatesArg {
  constructor(protected id: number, protected dates: Date[]) {
  }

  toRequestOptions(): IIdAndDates {
    const { id, dates } = this;

    const datesAsStrings = dates.map(date => toDateString(date));

    return {
      id,
      dates: datesAsStrings,
    };
  }
}

export const idDatesArg = (id: number, dates: Date[]): IIdAndDates => new IDAndDatesArg(
  id,
  dates,
).toRequestOptions();
