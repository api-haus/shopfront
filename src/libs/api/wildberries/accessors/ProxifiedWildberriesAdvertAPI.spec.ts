import { UTCDate } from '@date-fns/utc';
import { expect, test } from '@jest/globals';

import { testAdvertIds } from '../../../test/fixtures/advertIDs.js';
import { testMongoClient } from '../../../test/fixtures/testMongoClient.js';
import {
  testWbTokenProvider,
} from '../../../test/fixtures/testWbTokenProvider.js';
import { WildberriesAdvertAPI } from '../clients/WildberriesAdvertAPI.js';
import { idDatesArg } from '../params/IDAndDatesArg.js';
import { CachedWildberriesAdvertAPI } from './CachedWildberriesAdvertAPI.js';

test(
  'Thingy',
  async () => {
    const wbAdvertsAPI = new WildberriesAdvertAPI(testWbTokenProvider);
    const p = new CachedWildberriesAdvertAPI(
      wbAdvertsAPI,
      testMongoClient,
    );

    const allStats = await p.promotionFullStats([
      idDatesArg(
        testAdvertIds[45],
        [new UTCDate('2024-06-21')],
      ),
    ]);

    expect(allStats).toMatchSnapshot();
  },
);
