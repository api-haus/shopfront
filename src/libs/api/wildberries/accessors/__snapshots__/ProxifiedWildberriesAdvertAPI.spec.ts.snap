// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Thingy 1`] = `
[
  {
    "advertId": 17621826,
    "atbs": 1,
    "clicks": 82,
    "cpc": 16.57,
    "cr": 0,
    "ctr": 3.11,
    "dates": [
      "2024-06-24",
    ],
    "days": [
      {
        "apps": [
          {
            "appType": 1,
            "atbs": 0,
            "clicks": 4,
            "cpc": 50.38,
            "cr": 0,
            "ctr": 1,
            "nm": [
              {
                "atbs": 0,
                "clicks": 4,
                "cpc": 50.38,
                "cr": 0,
                "ctr": 1,
                "name": "Искусственное декоративное дерево Финиковая Пальма 220 см",
                "nmId": 231130515,
                "orders": 0,
                "shks": 0,
                "sum": 201.5,
                "sum_price": 0,
                "views": 399,
              },
            ],
            "orders": 0,
            "shks": 0,
            "sum": 201.5,
            "sum_price": 0,
            "views": 399,
          },
          {
            "appType": 32,
            "atbs": 0,
            "clicks": 40,
            "cpc": 19.14,
            "cr": 0,
            "ctr": 2.68,
            "nm": [
              {
                "atbs": 0,
                "clicks": 40,
                "cpc": 19.14,
                "cr": 0,
                "ctr": 2.68,
                "name": "Искусственное декоративное дерево Финиковая Пальма 220 см",
                "nmId": 231130515,
                "orders": 0,
                "shks": 0,
                "sum": 765.5,
                "sum_price": 0,
                "views": 1491,
              },
            ],
            "orders": 0,
            "shks": 0,
            "sum": 765.5,
            "sum_price": 0,
            "views": 1491,
          },
          {
            "appType": 64,
            "atbs": 1,
            "clicks": 38,
            "cpc": 0.01,
            "cr": 0,
            "ctr": 5.1,
            "nm": [
              {
                "atbs": 1,
                "clicks": 38,
                "cpc": 10.3,
                "cr": 0,
                "ctr": 5.1,
                "name": "Искусственное декоративное дерево Финиковая Пальма 220 см",
                "nmId": 231130515,
                "orders": 0,
                "shks": 0,
                "sum": 391.5,
                "sum_price": 0,
                "views": 745,
              },
            ],
            "orders": 0,
            "shks": 0,
            "sum": 0.39,
            "sum_price": 0,
            "views": 745,
          },
        ],
        "atbs": 1,
        "clicks": 82,
        "cpc": 16.57,
        "cr": 0,
        "ctr": 0,
        "date": "2024-06-24T03:00:00+03:00",
        "orders": 0,
        "shks": 0,
        "sum": 1358.5,
        "sum_price": 0,
        "views": 2635,
      },
    ],
    "orders": 0,
    "shks": 0,
    "sum": 1358.5,
    "sum_price": 0,
    "views": 2635,
  },
]
`;
